### PR TITLE
[CL]: Fix tick rounding bug and implement direct conversion from tick <> sqrt price

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #5516

> **Note to reviewer:** the only real state machine change is in `tick.go` and heavily mirrors @ValarDragon's PR here: #5522 
> The rest of the changes are related to function renames/test refactors.

## What is the purpose of the change

This PR expands on #5522 and updates all price to tick conversions to use SqrtPriceToTick instead.

## Testing and Verifying

The new function is tested in `math/tick_test.go`

The original attack vector is also converted into an edge case test in `position_test.go`

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/keeper_test.go
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/keeper_test.go:55]: Driveby rename as the original name could be misinterpreted to refer to rounding behavior

File: x/concentrated-liquidity/swaps.go
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:368]: Unsure whether in/out of the scope of this PR but we should not be doing tick spacing rounding in swaps, only in LP logic to prevent initialization of ticks we do not desire.

It should be possible to be within tick spacing for the "current tick" upon swap completion. However, it should not be possible to initialize the tick at an index that is not a multiple of tick spacing.

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/swaps.go:368]: Hmm thanks for flagging this – not sure I fully follow the concern but sounds important. Let's discuss separately

File: x/concentrated-liquidity/position_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/position_test.go:2535]: Should we assert any invariance here on the amount withdrawn?

File: x/concentrated-liquidity/range_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/range_test.go:153]: Do we need to add the pool's incentive address here as well?

File: x/concentrated-liquidity/swaps.go
Comment (czarcas7ic) [reply x/concentrated-liquidity/swaps.go:368]: Not sure if I am following here, do you mean a pools current tick inter swap can be between tick spacings or the result can be between tick spacings? I would think at the very least the resulting tick once the swap is complete should be rounded.

**Review by czarcas7ic — COMMENTED** [review]
Well done on this, will approve after some discourse in some of these comments

File: x/concentrated-liquidity/range_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/range_test.go:153]: I don't believe so – the reason we have to add spread rewards is because the tracking we do wrt swaps is inclusive of fees (we add `tokenIn` and subtract `tokenOut` from `totalAssets` after each swap). Incentives aren't part of this equation

File: x/concentrated-liquidity/position_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/position_test.go:2535]: We already cover this in the invariant fuzz tests so this behavior should already be thoroughly covered. I mostly just added this case as a formality to show that the original attack vector no longer triggered a panic

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:271]: It's minus 2 because to check tick t + 1, we need to go to t + 2. Thus, to not go over max tick during these checks, we need to shift down by 2

File: x/concentrated-liquidity/invariant_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/invariant_test.go:150]: This is a good point. Investigating now.

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/invariant_test.go]: Discussed offline and converged on relaxing the invariant to simply check for rounding direction instead of specifying an exact additive amount, as we slightly overround in parts of our logic and it is not worth blocking this PR on changing this

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:278]: Err might be missing your point – if truncated tick = MaxTick, it enters this branch since MaxTick >= MaxTick - 1, no?

Comment (czarcas7ic) [reply x/concentrated-liquidity/math/tick.go:278]: You are right, was just not clear on first pass, thanks

**Review by czarcas7ic — APPROVED** [review]
Spoke offline IRT that unanswered questions. Approving but looking for a second approval before merge. Nice work again :)

File: x/concentrated-liquidity/swaps.go
Comment (p0mvn) [reply x/concentrated-liquidity/swaps.go:368]: Yes, that's what I mean. Current tick can be between two initialized ticks inter-swap. That is, inside an active bucket. With tick rounding, the translation between current sqrt price and current tick would be broken.

I addressed this concern in this PR: https://github.com/osmosis-labs/osmosis/pull/5526 and left more detailed explanation.

Let's continue discussing this in #5526

File: x/concentrated-liquidity/position_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/position_test.go:2535]: Commenting for visibility, I expect this to solve this issue in #5526 : 
https://github.com/osmosis-labs/osmosis/pull/5526/files#r1233117388

File: x/concentrated-liquidity/math/tick.go
Comment (p0mvn) [inline x/concentrated-liquidity/math/tick.go:256]: Per definitions here:
https://github.com/osmosis-labs/osmosis/blob/bd809a973132d55a63e7c34c326ec1408b38eec2/x/concentrated-liquidity/README.md?plain=1#L1589-L1596

I would not call this a bucket but, instead, a tick. The reason why I think these semantics are important is that a tick can be within a bucket.

File: x/concentrated-liquidity/math/tick.go
Comment (p0mvn) [reply x/concentrated-liquidity/math/tick.go:150]: +1. If we are set on the new approach, I would like to suggest removing the old broken logic.

I understand that it will require refactoring tests and can be done in a separate PR. However, I think it would be a time well spent as a few months later the context will be lost, and someone will start wondering why we have two kinds of price to tick conversions.

For this PR, I think we should at least move it outside of the core logic like Adam suggested

File: x/concentrated-liquidity/math/tick.go
Comment (p0mvn) [reply x/concentrated-liquidity/math/tick.go:271]: Could you please add this context to the comments?

**Review by p0mvn — COMMENTED** [review]
Great work! I would like to do another pass tomorrow and run tests more. Leaving nits for now

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:256]: Hmm yeah it seems the definition has shifted slightly (this was a comment dev originally added). I just removed the note to avoid further confusion

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: Of course. Added here: 2a4f1c4cd6e3c4b14da70dceeb0fbf92eaa45faa

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: Discussed offline and decided to move into apptesting as a short term fix to get these functions out of the state machine: 08bd83f286f1ab2ecd2b493717b95569eb73acb1

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: 08bd83f286f1ab2ecd2b493717b95569eb73acb1

File: app/apptesting/concentrated_liquidity.go
Comment (AlpinYukseloglu) [inline app/apptesting/concentrated_liquidity.go:152]: Note to reviewer: these functions were moved to be high level test helpers to get them out of the state machine without breaking the many tests that rely on them to (correctly) generate expected values

File: x/concentrated-liquidity/range_test.go
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/range_test.go:50]: Note to reviewer: the changes in this file are mainly cleanup/test speed related changes to ensure tests run in a reasonable amount of time

File: x/concentrated-liquidity/math/tick.go
Comment (ValarDragon) [reply x/concentrated-liquidity/math/tick.go:242]: I am confused about the `!outOfBounds` conditional as tied to this

File: x/concentrated-liquidity/position_test.go
Comment (ValarDragon) [inline x/concentrated-liquidity/position_test.go:2558]: does this case still error after this PR?

File: x/concentrated-liquidity/math/tick.go
Comment (ValarDragon) [inline x/concentrated-liquidity/math/tick.go:224]: I wonder if its simpler to just deal with these by checking the boundary cases in a separate function, on the sqrt price directly. (pre-squaring)

File: x/concentrated-liquidity/math/tick_test.go
Comment (p0mvn) [inline x/concentrated-liquidity/math/tick_test.go:801]: Should we also be having test cases around min tick bounds similar to how this is done for max tick?

File: x/concentrated-liquidity/position_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/position_test.go:2558]: No since we relaxed the invariant checks to allow for slight overrounding recently. Will uncomment

**Review by p0mvn — APPROVED** [review]
LGTM. Great work

File: x/concentrated-liquidity/math/tick.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:224]: Possibly, although my previous attempts at this led to constraining the "real" min tick and max tick by 1-2 ticks (which might be fine but leads to more cases that need to be checked/handled)

If you see a way to abstract this logic without triggering this, would be great for us to do a follow up PR to clean this up!

File: x/concentrated-liquidity/math/tick_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick_test.go:801]: Will add these cases in the min tick change PR as the expected values are thrown off due to canonical tick rounding

**Pull request merged**
